src/Adapter/MSTest.TestAdapter/Execution/TestClassInfo.cs

```diff
@@ -383,56 +405,57 @@ public void RunClassInitialize(TestContext testContext)
 
             if (IsClassInitializeExecuted || ClassInitializeMethod is null)
             {
-                MethodInfo? classCleanupMethod = null;
-
                 try
                 {
                     classCleanupMethod = ClassCleanupMethod;
-                    classCleanupMethod?.InvokeAsSynchronousTask(null);
+                    ClassCleanupException = InvokeCleanupMethod(classCleanupMethod);
                     var baseClassCleanupQueue = new Queue<MethodInfo>(BaseClassCleanupMethodsStack);
-                    while (baseClassCleanupQueue.Count > 0)
+                    while (baseClassCleanupQueue.Count > 0 && ClassCleanupException is null)
                     {
                         classCleanupMethod = baseClassCleanupQueue.Dequeue();
-                        classCleanupMethod?.InvokeAsSynchronousTask(null);
+                        ClassCleanupException = InvokeCleanupMethod(classCleanupMethod);
                     }
-
-                    IsClassCleanupExecuted = true;
-
-                    return null;
                 }
                 catch (Exception exception)
                 {
-                    var realException = exception.GetRealException();
-                    ClassCleanupException = realException;
+                    ClassCleanupException = exception;
+                }
+            }
+        }
 
-                    // special case AssertFailedException to trim off part of the stack trace
-                    string errorMessage = realException is AssertFailedException or AssertInconclusiveException
-                        ? realException.Message
-                        : realException.GetFormattedExceptionMessage();
+        // If ClassCleanup was successful, then don't do anything
+        if (ClassCleanupException == null)
+        {
+            // We didn't but it in a finally function above to not change the old behavior as it was only in the the try
```

Comment:
I don't really get this message, could you give me a little more insights?

Reply:
comment is related to that line
https://github.com/microsoft/testfx/pull/2570/files/57d6b5a709c2220520229da2b541f22028e07a82#diff-894e540fd1b1ac62982e36ac99212fcf3c5d0123db60951b3809b5c4f569717cR430

as old code has this line inside the try so if exception happens that flag would stay false 
so I didn't add it in `finally`

Reply:
This is confusing also to me, I would keep the set at line 399 so the semantic is clear, if we don't have exception we run all the init, move down here "far" for the invocation could confuse the reader. The fact that we evaluate the exception here is only because we can set it in 2 place inside try/catch(411) and inside the block catch(421).

Reply:
I could have exception inside the try assigned to the exception var, so I can't set inside the try until i check the var
what do you think?